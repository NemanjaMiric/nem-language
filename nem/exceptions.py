"""Raise exceptions.

Holds multiple exceptions.

Exceptions:
LexerException - During lexing
ParserException - During parsing
EvaluatorException - During evaluation

"""


class LexerException(Exception):

    """Raise exception.

    Gets raised when an error appears during lexing.

    """

    pass


class ParserException(Exception):

    """Raise exception.

    Gets raised when an error appears during parsing.

    """

    pass


class EvaluatorException(Exception):

    """Raise exception.

    Gets raised when an error appears during evaluation.

    """

    pass