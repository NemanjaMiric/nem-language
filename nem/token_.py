"""Represent a token.

Holds class Token which is used when the code is being lexed.

"""


class Token:

    """Represent a token.

    Used for representing a token when the code is being lexed.

    """

    # Tokens
    NUMBER = "NUMBER"
    TEXT = "TEXT"
    SYMBOL = "SYMBOL"
    NULL = "NULL"
    COMMA = "COMMA"
    CARET = "CARET"
    ASTERISK = "ASTERISK"
    SLASH = "SLASH"
    PLUS = "PLUS"
    MINUS = "MINUS"
    EQUAL = "EQUAL"
    WHILE = "WHILE"
    LEFT_BRACKET = "LEFT_BRACKET"
    RIGHT_BRACKET = "RIGHT_BRACKET"
    IF = "IF"
    OTHERWISE = "OTHERWISE"
    FUNCTION = "FUNCTION"
    IMPORT = "IMPORT"
    LEFT_SQUARE = "LEFT_SQUARE"
    RIGHT_SQUARE = "RIGHT_SQUARE"
    BREAK = "BREAK"
    CONTINUE = "CONTINUE"
    RETURN = "RETURN"
    NOT = "NOT"
    OR = "OR"
    AND = "AND"
    LESS = "LESS"
    LESS_EQUAL = "LESS_EQUAL"
    GREATER = "GREATER"
    GREATER_EQUAL = "GREATER_EQUAL"
    IS = "IS"
    MODULO = "MODULO"
    EOF = "EOF"

    def __init__(self, type_, value, filename, line):
        """Initialize class Token.

        Used for initializing the Token class.

        """
        self.type = type_
        self.value = value
        self.filename = filename
        self.line = line

    def __repr__(self):
        """Represent Token class."""
        return "Token({}, {}, {}, {})".format(repr(self.type), repr(self.value), repr(self.filename), repr(self.line))

    def __eq__(self, other):
        """Check equality of Token class."""
        if isinstance(other, Token):
            return self.type == other.type and self.value == other.value and self.line == other.line
        else:
            raise NotImplementedError
