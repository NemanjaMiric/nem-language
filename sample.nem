# Fibonacci numbers generator

function fib(first, second, stop) (
    while (first < stop) (
        temporary = first
        first = second
        second = temporary + second
    )

    return 0
)

fib(1, 1, 1000)

# Higher-order function combined with function list indexing

function hof() (
    function inner_func() (
        return [[1, 2], 2]
    )
    return inner_func
)

hof()()[0][1]
