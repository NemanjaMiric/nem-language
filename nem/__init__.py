"""Initialize modules."""

import nem.evaluator
import nem.exceptions
import nem.interpreter
import nem.lexer
import nem.nodes
import nem.parser
import nem.symbol_table
import nem.token_
import nem.types_
