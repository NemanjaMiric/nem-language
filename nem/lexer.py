"""Hold Lexer class.

Holds the Lexer class which is used for converting raw source code into tokens.

Tokens (Extended Backus-Naur form):
    LETTER_           = "A" | "B" | "C" | "D" | "E" | "F" | "G"
                      | "H" | "I" | "J" | "K" | "L" | "M" | "N"
                      | "O" | "P" | "Q" | "R" | "S" | "T" | "U"
                      | "V" | "W" | "X" | "Y" | "Z" | "a" | "b"
                      | "c" | "d" | "e" | "f" | "g" | "h" | "i"
                      | "j" | "k" | "l" | "m" | "n" | "o" | "p"
                      | "q" | "r" | "s" | "t" | "u" | "v" | "w"
                      | "x" | "y" | "z" ;
    DIGIT_            = "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" ;
    ESCAPED_          = "\n" | "\t" | "\\" | '\"' ;
    NUMBER            = { DIGIT_ }, [ "." ], DIGIT_, { DIGIT_ } ;
    TEXT              = '"', { ? any character other than double quotation mark ? | ESCAPED_ }, '"' ;
    SYMBOL            = LETTER_ | "_", { LETTER_ | "_" | DIGIT_ } ;
    NULL              = "null" ;
    COMMA             = "," ;
    CARET             = "^" ;
    ASTERISK          = "*" ;
    SLASH             = "/" ;
    PLUS              = "+" ;
    MINUS             = "-" ;
    EQUAL             = "=" ;
    WHILE             = "while" ;
    LEFT_BRACKET      = "(" ;
    RIGHT_BRACKET     = ")" ;
    IF                = "if" ;
    OTHERWISE         = "otherwise" ;
    FUNCTION          = "function" ;
    IMPORT            = "import" ;
    LEFT_SQUARE       = "[" ;
    RIGHT_SQUARE      = "]" ;
    BREAK             = "break" ;
    CONTINUE          = "continue" ;
    RETURN            = "return" ;
    NOT               = "not" ;
    OR                = "or" ;
    AND               = "and" ;
    LESS              = "<" ;
    LESS_EQUAL        = "<=" ;
    GREATER           = ">" ;
    GREATER_EQUAL     = ">=" ;
    IS                = "is" ;
    MODULO            = "%" ;
    EOF               = ? end of file ? ;
    COMMENT_          = "#", { ? any character other than newline ? }, "\n" ;

"""

from nem.token_ import Token
from nem.exceptions import LexerException
import re


class Lexer:

    """Convert source code into tokens.

    Used for converting raw source code into tokens.

    """

    SYMBOLS = {
        "null": Token.NULL,
        "while": Token.WHILE,
        "if": Token.IF,
        "otherwise": Token.OTHERWISE,
        "function": Token.FUNCTION,
        "import": Token.IMPORT,
        "break": Token.BREAK,
        "continue": Token.CONTINUE,
        "return": Token.RETURN,
        "not": Token.NOT,
        "or": Token.OR,
        "and": Token.AND,
        "is": Token.IS
    }

    def __init__(self, code, filename):
        """Initialize Lexer class."""
        self.code = code
        self.filename = filename

        self.current_index = 0
        self.previous_character = None
        self.current_character = code[0]

        self.line = 1

    def _advance_index(self):
        """Advance the current character."""
        self.current_index += 1
        self.previous_character = self.current_character
        self.current_character = self.code[self.current_index] if self.current_index < len(self.code) else None
        if self.previous_character == "\n" and self.current_character is not None:
            self.line += 1

    def _next_number(self):
        """Return next number relative to the current position."""
        # Tries to find the next number
        while self.current_character is not None and not re.match("[0-9.]", self.current_character):
            self._advance_index()

        if self.current_character is None:
            raise LexerException("Lexing Error (File {}) (Line {}): No number found.".format(self.filename, self.line))

        length = 0

        # "[0-9]*\.?[0-9]*\Z" checks for a valid number (10, 10.2, .4, etc.)
        while self.current_character is not None                                                                       \
                and re.match("[0-9]*\.?[0-9]+\Z", self.code[self.current_index - length:self.current_index + 1])       \
                or self.current_character == ".":
            length += 1
            self._advance_index()

        if self.previous_character == ".":
            raise LexerException("Lexing Error (File {}) (Line {}): Unrecognized token"
                                 .format(self.filename, self.line))

        return self.code[self.current_index - length:self.current_index]

    def _next_symbol(self):
        """Return next symbol."""
        # Tries to find the next symbol
        while self.current_character is not None and not re.match("[a-zA-Z_]", self.current_character):
            self._advance_index()

        if self.current_character is None:
            raise LexerException("Lexing Error (File {}) (Line {}): No symbol found.".format(self.filename, self.line))

        length = 0

        # "[a-zA-Z_][a-zA-Z_0-9]*\Z" checks for a valid symbol, but many other tokens (purposefully) pass as well
        while self.current_character is not None                                                                       \
                and re.match("[a-zA-Z_][a-zA-Z_0-9]*\Z", self.code[self.current_index - length:self.current_index + 1]):
            length += 1
            self._advance_index()

        return self.code[self.current_index - length:self.current_index]

    def _next_text(self):
        """Return next text.

        Everything between the current or found quotation mark and the next unescaped one gets lexed.

        """
        # Tries to find the next text
        while self.current_character is not None and self.current_character != "\"":
            self._advance_index()

        if self.current_character is None:
            raise LexerException("Lexing Error (File {}) (Line {}): No text found.".format(self.filename, self.line))

        length = 0
        self._advance_index()

        escaped_characters = (
            "n",
            "t",
            "\"",
            "\\"
        )

        # Up until it reaches the closing quotation mark, it continues adding all the characters to text
        while self.current_character is not None and self.current_character != "\"":
            length += 1
            self._advance_index()

            if self.previous_character == "\\" and self.current_character in escaped_characters:
                length += 1
                self._advance_index()

        if self.current_character is None:
            raise LexerException("Lexing Error (File {}) (Line {}): Text doesn't have an end"
                                 .format(self.filename, self.line))

        self._advance_index()

        # -1s are added because of self._advance_index() at the end, which increases self.current_index by 1
        return self.code[self.current_index - length - 1:self.current_index - 1]\
            .replace("\\n", "\n").replace("\\t", "\t").replace('\\"', "\"")

    def _next_token(self):
        """Return next token."""
        if self.current_character is None:
            token = None

        # Ignores whitespaces and tabs
        elif self.current_character in " \t":
            self._advance_index()
            token = False

        # Lexes COMMA token
        elif self.current_character == ",":
            token = Token(Token.COMMA, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes CARET token
        elif self.current_character == "^":
            token = Token(Token.CARET, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes ASTERISK token
        elif self.current_character == "*":
            token = Token(Token.ASTERISK, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes SLASH token
        elif self.current_character == "/":
            token = Token(Token.SLASH, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes PLUS token
        elif self.current_character == "+":
            token = Token(Token.PLUS, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes MINUS token
        elif self.current_character == "-":
            token = Token(Token.MINUS, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes MODULO token
        elif self.current_character == "%":
            token = Token(Token.MODULO, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes EQUAL token
        elif self.current_character == "=":
            token = Token(Token.EQUAL, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes LEFT_BRACKET token
        elif self.current_character == "(":
            token = Token(Token.LEFT_BRACKET, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes RIGHT_BRACKET token
        elif self.current_character == ")":
            token = Token(Token.RIGHT_BRACKET, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes LEFT_SQUARE token
        elif self.current_character == "[":
            token = Token(Token.LEFT_SQUARE, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes RIGHT_SQUARE token
        elif self.current_character == "]":
            token = Token(Token.RIGHT_SQUARE, self.current_character, self.filename, self.line)
            self._advance_index()

        # Lexes LESS or LESS_EQUAL token
        elif self.current_character == "<":
            self._advance_index()
            if self.current_character == "=":
                token = Token(Token.LESS_EQUAL, "{}{}".format(self.previous_character, self.current_character),
                              self.filename, self.line)
                self._advance_index()
            else:
                token = Token(Token.LESS, self.previous_character, self.filename, self.line)

        # Lexes GREATER or GREATER_EQUAL token
        elif self.current_character == ">":
            self._advance_index()
            if self.current_character == "=":
                token = Token(Token.GREATER_EQUAL, "{}{}".format(self.previous_character, self.current_character),
                              self.filename, self.line)
                self._advance_index()
            else:
                token = Token(Token.GREATER, self.previous_character, self.filename, self.line)

        # Lexes TEXT token - everything between the current quotation mark and the next one gets lexed
        elif self.current_character == "\"":
            token = Token(Token.TEXT, self._next_text(), self.filename, self.line)

        # Lexes NUMBER token
        elif re.match("[0-9.]", self.current_character):
            token = Token(Token.NUMBER, self._next_number(), self.filename, self.line)

        # Lexes both SYMBOL tokens and tokens comprised of letters
        elif re.match("[a-zA-Z_]", self.current_character):
            temporary = self._next_symbol()

            try:
                token = Token(self.SYMBOLS[temporary], temporary, self.filename, self.line)
            except KeyError:
                # If it's not a token comprised of letters, it's a SYMBOL token
                token = Token(Token.SYMBOL, temporary, self.filename, self.line)

        elif self.current_character == "\n":
            self._advance_index()
            token = False

        # Comments - everything in-between '#' character and the next newline character gets ignored
        elif self.current_character == "#":
            while self.current_character is not None and self.current_character != "\n":
                self._advance_index()
            self._advance_index()

            token = False

        else:
            raise LexerException("Lexing Error (File {}) (Line {}): Unrecognized token"
                                 .format(self.filename, self.line))

        return token

    def _tokens(self):
        """Return all tokens."""
        current_token = self._next_token()

        while current_token is not None:
            # If the current token is a comment, skips the yield
            if current_token:
                yield current_token
            current_token = self._next_token()

        # Appends the EOF token
        yield Token(Token.EOF, None, self.filename, self.line)

    def lex(self):
        """Lex the code.

        Lexer for Nem code.

        Tokens (Extended Backus-Naur form):
            LETTER_           = "A" | "B" | "C" | "D" | "E" | "F" | "G"
                              | "H" | "I" | "J" | "K" | "L" | "M" | "N"
                              | "O" | "P" | "Q" | "R" | "S" | "T" | "U"
                              | "V" | "W" | "X" | "Y" | "Z" | "a" | "b"
                              | "c" | "d" | "e" | "f" | "g" | "h" | "i"
                              | "j" | "k" | "l" | "m" | "n" | "o" | "p"
                              | "q" | "r" | "s" | "t" | "u" | "v" | "w"
                              | "x" | "y" | "z" ;
            DIGIT_            = "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" ;
            ESCAPED_          = "\n" | "\t" | "\\" | '\"' ;
            NUMBER            = { DIGIT_ }, [ "." ], DIGIT_, { DIGIT_ } ;
            TEXT              = '"', { ? any character other than double quotation mark ? | ESCAPED_ }, '"' ;
            SYMBOL            = LETTER_ | "_", { LETTER_ | "_" | DIGIT_ } ;
            NULL              = "null" ;
            COMMA             = "," ;
            CARET             = "^" ;
            ASTERISK          = "*" ;
            SLASH             = "/" ;
            PLUS              = "+" ;
            MINUS             = "-" ;
            EQUAL             = "=" ;
            WHILE             = "while" ;
            LEFT_BRACKET      = "(" ;
            RIGHT_BRACKET     = ")" ;
            IF                = "if" ;
            OTHERWISE         = "otherwise" ;
            FUNCTION          = "function" ;
            IMPORT            = "import" ;
            LEFT_SQUARE       = "[" ;
            RIGHT_SQUARE      = "]" ;
            BREAK             = "break" ;
            CONTINUE          = "continue" ;
            RETURN            = "return" ;
            NOT               = "not" ;
            OR                = "or" ;
            AND               = "and" ;
            LESS              = "<" ;
            LESS_EQUAL        = "<=" ;
            GREATER           = ">" ;
            GREATER_EQUAL     = ">=" ;
            IS                = "is" ;
            MODULO            = "%" ;
            EOF               = ? end of file ? ;
            COMMENT_          = "#", { ? any character other than newline ? }, "\n" ;

        """
        return self._tokens()


def main():
    """Debug the lexer.

    Used as the entry-point when the file gets ran directly. Used for debugging the class Lexer.

    """
    while True:
        print(list(Lexer(input(">> ") + "\n", "<stdin>").lex()))


if __name__ == "__main__":
    # Only activates when the file gets ran directly.
    main()
