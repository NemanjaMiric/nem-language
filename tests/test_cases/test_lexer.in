# Test comment

# Test comment
VAriable_name = .2
variable_name = 10.2 # Test comment
variable_name_2 = "a string" # Test comment


# Test comment
10 + 2.2 # Test comment
22 + "17" # Test comment
22 + "a" # Test comment
23 + 1, 2, 3, # Test comment
2 + null # Test comment

"22" + 23 # Test comment
"asf" + "ghk" # Test comment
"df" + 1, 2, 3, # Test comment
"c" + null # Test comment

1, 2, 3, + 6 # Test comment
1, 2, 3, + "2f" # Test comment
1, 2, 3, + 2, 3, # Test comment
1, 2, 3, + null # Test comment

null+1 # Test comment
null+"2" # Test comment
null+"a" # Test comment
null+1, 2, 3, # Test comment
null+null # Test comment


# Test comment
2 - 7 # Test comment
3 - "1" # Test comment
3 - "a" # Test comment
5 - 1, 2, 3, # Test comment
2 - 1, 2, 3, 5, # Test comment
2 - null # Test comment

"a3t42847766t" - 76 # Test comment
"sdf" - "f" # Test comment
"fff" - 1, 2, "f", # Test comment
"asf" - null # Test comment

1, 2, 3, 2, - 2 # Test comment
1, 2, - "2" # Test comment
2, 3, - 3, 3, # Test comment
1, 3, - null # Test comment

null-1 # Test comment
null-"2" # Test comment
null-"a" # Test comment
null-1, 2, # Test comment
null-null # Test comment


# Test comment
2 * 2 # Test comment
13 * "2" # Test comment
13 * "a" # Test comment
2 * 1, 2, # Test comment
3 * null # Test comment

"21" * 6 # Test comment
"3" * "5" # Test comment
"a" * "5" # Test comment
"dd" * "as" # Test comment
"2" * 1, 2, # Test comment
"a" * 1, 2, # Test comment
"asd" * null # Test comment

1, 2, * 2 # Test comment
1, 2, * "3" # Test comment
1, 2, * "as2d" # Test comment
1, 2, * 1, 2, 3, # Test comment
1, 3, * null # Test comment

null*2 # Test comment
null*"3" # Test comment
null*"a" # Test comment
null*1,2, # Test comment
null*null # Test comment


# Test comment
5 / 2 # Test comment
3 / "1" # Test comment
2 / "a" # Test comment
5 / 1, 2, # Test comment
4 / null # Test comment
10 / 0 # Test comment

"a" / 2 # Test comment
"5" / 2 # Test comment
"5" / "2" # Test comment
"25" / 1, 2, # Test comment
"f2" / null # Test comment
"4" / "0" # Test comment

1, 2, / 1 # Test comment
1, / 2 # Test comment
2, 3, 4, / "a" # Test comment
2, 3 / 2, # Test comment
2, 3, / null # Test comment

null/4 # Test comment
null/".2" # Test comment
null/"a" # Test comment
null/1, 2, # Test comment
null/null # Test comment


# Test comment
2 ^ 3 # Test comment
2 ^ "2" # Test comment
2 ^ "a" # Test comment
3 ^ 1, 2, # Test comment
2 ^ null # Test comment

"a" ^ 2 # Test comment
"2" ^ 3 # Test comment
"df" ^ "a" # Test comment
"asd" ^ 1, 2, # Test comment
"asd" ^ null # Test comment

1, 2, ^ 1 # Test comment
1, 2, ^ "a" # Test comment
1, 2, ^ 1, # Test comment
1, 2, ^ null # Test comment

null^1 # Test comment
null^"2" # Test comment
null^1,2, # Test comment
null^null # Test comment


# Test comment
5 % 2 # Test comment
5 % "2" # Test comment
5 % "a" # Test comment
3 % 1, 2, "df", # Test comment
2 % null # Test comment
3 % 0 # Test comment
2 % "0" # Test comment

"a" % 2 # Test comment
"2" % 2 # Test comment
"d" % "a" # Test comment
"2" % "1" # Test comment
"asd" % 1, 2, # Test comment
"asd" % null # Test comment

1, 2, % 1 # Test comment
1, 2, % "a" # Test comment
1, 2, % 1, 2, # Test comment
1, 2, % null # Test comment

null%1 # Test comment
null%"2" # Test comment
null%"a" # Test comment
null%1,2, # Test comment
null%null # Test comment


# Test comment
input variable_name # Test comment


# Test comment
output variable_name # Test comment


# Test comment
error variable_name # Test comment


# Test comment
condition = 1
while condition (condition = 0) # Test comment


# Test comment
if condition (a=2) otherwise (a=3) # Test comment
if condition (a=2) # Test comment


# Test comment
function function_name(parameter_1, parameter_2) (parameter_1 = parameter_2 return parameter1) # Test comment

function_name(1, 2) # Test comment

# Test comment
import sample # Test comment
sample_function_name(2, 2) # Test comment


# Test comment
file "README.md" read variable_name # Test comment
file "README.md" write variable_name # Test comment


# Test comment
convert variable_name number # Test comment
convert variable_name text # Test comment
convert variable_name list # Test comment


# Test comment
a_list = 1, 2, 3, "a",
a_list[0] # Test comment


# Test comment
i = 0
while (i < 100)
(
    i = i + 1
    if (i > 20) (
        break
    )
    else (
        output i
        continue
    )
)


# Test comment
if (not 1) (a=2)
if (not "") (b=2)


# Test comment
if (1 is 2) (a=2)
if (1 is 1) (a=3)


# Test comment
if (1 is 1 and 2 is 2) (a=2)


# Test comment
if (1 is 2 or 1 is 1) (a=2)


# Test comment
if (not (1 > 2)) (a = 2)

# Test comment
if (1 >= 2) (b = -1)

# Test comment
if (1 < 2) (a = 2)

# Test comment
if (2 <= 2) (b = "as")