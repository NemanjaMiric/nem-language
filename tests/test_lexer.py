"""Test for Lexer class.

Unit testing for the Lexer class.

"""


import unittest
from nem.lexer import Lexer
from nem.token_ import Token
from nem.exceptions import LexerException


class LexerTestCase(unittest.TestCase):

    """Unit test Lexer class.

    Used for unit testing the Lexer class.

    """

    def test_lex(self):
        """Test the lex method.

        Tests the lexer.

        """
        with open("test_cases/test_lexer.in") as _input:
            _input = _input.read()

        with open("test_cases/test_lexer.out") as output:
            output = eval(output.read())

        self.assertEqual(list(Lexer(_input).lex()), output)

        with self.assertRaises(LexerException):
            list(Lexer("\n\"\"\"").lex())

        with self.assertRaises(LexerException):
            list(Lexer("\n..\n").lex())

        with self.assertRaises(LexerException):
            list(Lexer(".\n").lex())

        with self.assertRaises(LexerException):
            list(Lexer("\n'\n").lex())


if __name__ == '__main__':
    # Only activates when the file gets ran directly.
    unittest.main()
