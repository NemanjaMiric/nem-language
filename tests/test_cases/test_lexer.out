[Token('IMPORT', 'import', '<stdin>', 3), Token('TEXT', 'test', '<stdin>', 3), Token('FUNCTION', 'function', '<stdin>', 5), Token('SYMBOL', 'fib', '<stdin>', 5), Token('LEFT_BRACKET', '(', '<stdin>', 5), Token('SYMBOL', 'first', '<stdin>', 5), Token('COMMA', ',', '<stdin>', 5), Token('SYMBOL', 'second', '<stdin>', 5), Token('COMMA', ',', '<stdin>', 5), Token('SYMBOL', 'stop', '<stdin>', 5), Token('RIGHT_BRACKET', ')', '<stdin>', 5), Token('LEFT_BRACKET', '(', '<stdin>', 5), Token('WHILE', 'while', '<stdin>', 6), Token('LEFT_BRACKET', '(', '<stdin>', 6), Token('NOT', 'not', '<stdin>', 6), Token('NUMBER', '0', '<stdin>', 6), Token('AND', 'and', '<stdin>', 6), Token('NOT', 'not', '<stdin>', 6), Token('NUMBER', '0', '<stdin>', 6), Token('RIGHT_BRACKET', ')', '<stdin>', 6), Token('LEFT_BRACKET', '(', '<stdin>', 6), Token('IF', 'if', '<stdin>', 7), Token('LEFT_BRACKET', '(', '<stdin>', 7), Token('SYMBOL', 'first', '<stdin>', 7), Token('LESS', '<', '<stdin>', 7), Token('SYMBOL', 'stop', '<stdin>', 7), Token('RIGHT_BRACKET', ')', '<stdin>', 7), Token('LEFT_BRACKET', '(', '<stdin>', 7), Token('SYMBOL', 'temporary', '<stdin>', 8), Token('EQUAL', '=', '<stdin>', 8), Token('SYMBOL', 'first', '<stdin>', 8), Token('SYMBOL', 'first', '<stdin>', 9), Token('EQUAL', '=', '<stdin>', 9), Token('SYMBOL', 'second', '<stdin>', 9), Token('SYMBOL', 'second', '<stdin>', 10), Token('EQUAL', '=', '<stdin>', 10), Token('SYMBOL', 'temporary', '<stdin>', 10), Token('PLUS', '+', '<stdin>', 10), Token('SYMBOL', 'second', '<stdin>', 10), Token('RIGHT_BRACKET', ')', '<stdin>', 11), Token('OTHERWISE', 'otherwise', '<stdin>', 11), Token('BREAK', 'break', '<stdin>', 11), Token('CONTINUE', 'continue', '<stdin>', 12), Token('RIGHT_BRACKET', ')', '<stdin>', 13), Token('RETURN', 'return', '<stdin>', 15), Token('NUMBER', '0', '<stdin>', 15), Token('RIGHT_BRACKET', ')', '<stdin>', 16), Token('SYMBOL', 'fib', '<stdin>', 18), Token('LEFT_BRACKET', '(', '<stdin>', 18), Token('NUMBER', '1', '<stdin>', 18), Token('COMMA', ',', '<stdin>', 18), Token('NUMBER', '1', '<stdin>', 18), Token('COMMA', ',', '<stdin>', 18), Token('NUMBER', '1000', '<stdin>', 18), Token('RIGHT_BRACKET', ')', '<stdin>', 18), Token('SYMBOL', 'a1', '<stdin>', 20), Token('EQUAL', '=', '<stdin>', 20), Token('LEFT_SQUARE', '[', '<stdin>', 20), Token('NUMBER', '1', '<stdin>', 20), Token('COMMA', ',', '<stdin>', 20), Token('NULL', 'null', '<stdin>', 20), Token('COMMA', ',', '<stdin>', 20), Token('TEXT', '"text"', '<stdin>', 20), Token('RIGHT_SQUARE', ']', '<stdin>', 20), Token('SYMBOL', '_0', '<stdin>', 21), Token('EQUAL', '=', '<stdin>', 21), Token('LEFT_BRACKET', '(', '<stdin>', 21), Token('MINUS', '-', '<stdin>', 21), Token('NUMBER', '0.2', '<stdin>', 21), Token('PLUS', '+', '<stdin>', 21), Token('NUMBER', '.2', '<stdin>', 21), Token('RIGHT_BRACKET', ')', '<stdin>', 21), Token('CARET', '^', '<stdin>', 21), Token('LEFT_BRACKET', '(', '<stdin>', 21), Token('NUMBER', '2', '<stdin>', 21), Token('MODULO', '%', '<stdin>', 21), Token('NUMBER', '2', '<stdin>', 21), Token('RIGHT_BRACKET', ')', '<stdin>', 21), Token('ASTERISK', '*', '<stdin>', 21), Token('LEFT_BRACKET', '(', '<stdin>', 21), Token('MINUS', '-', '<stdin>', 21), Token('NUMBER', '2', '<stdin>', 21), Token('PLUS', '+', '<stdin>', 21), Token('NUMBER', '02', '<stdin>', 21), Token('RIGHT_BRACKET', ')', '<stdin>', 21), Token('CARET', '^', '<stdin>', 21), Token('LEFT_BRACKET', '(', '<stdin>', 21), Token('NUMBER', '2', '<stdin>', 21), Token('MINUS', '-', '<stdin>', 21), Token('NUMBER', '2', '<stdin>', 21), Token('RIGHT_BRACKET', ')', '<stdin>', 21), Token('PLUS', '+', '<stdin>', 21), Token('LEFT_BRACKET', '(', '<stdin>', 21), Token('MINUS', '-', '<stdin>', 21), Token('NUMBER', '2', '<stdin>', 21), Token('MINUS', '-', '<stdin>', 21), Token('NUMBER', '2', '<stdin>', 21), Token('RIGHT_BRACKET', ')', '<stdin>', 21), Token('CARET', '^', '<stdin>', 21), Token('LEFT_BRACKET', '(', '<stdin>', 21), Token('NUMBER', '2', '<stdin>', 21), Token('PLUS', '+', '<stdin>', 21), Token('NUMBER', '2', '<stdin>', 21), Token('RIGHT_BRACKET', ')', '<stdin>', 21), Token('SLASH', '/', '<stdin>', 21), Token('LEFT_BRACKET', '(', '<stdin>', 21), Token('NUMBER', '2', '<stdin>', 21), Token('PLUS', '+', '<stdin>', 21), Token('NUMBER', '2', '<stdin>', 21), Token('RIGHT_BRACKET', ')', '<stdin>', 21), Token('CARET', '^', '<stdin>', 21), Token('LEFT_BRACKET', '(', '<stdin>', 21), Token('NUMBER', '2', '<stdin>', 21), Token('PLUS', '+', '<stdin>', 21), Token('NUMBER', '2', '<stdin>', 21), Token('RIGHT_BRACKET', ')', '<stdin>', 21), Token('SYMBOL', 'b', '<stdin>', 23), Token('EQUAL', '=', '<stdin>', 23), Token('SYMBOL', 'a', '<stdin>', 23), Token('LEFT_SQUARE', '[', '<stdin>', 23), Token('NUMBER', '1', '<stdin>', 23), Token('RIGHT_SQUARE', ']', '<stdin>', 23), Token('FUNCTION', 'function', '<stdin>', 25), Token('SYMBOL', 'a', '<stdin>', 25), Token('LEFT_BRACKET', '(', '<stdin>', 25), Token('RIGHT_BRACKET', ')', '<stdin>', 25), Token('LEFT_BRACKET', '(', '<stdin>', 25), Token('FUNCTION', 'function', '<stdin>', 26), Token('SYMBOL', 'b', '<stdin>', 26), Token('LEFT_BRACKET', '(', '<stdin>', 26), Token('RIGHT_BRACKET', ')', '<stdin>', 26), Token('RETURN', 'return', '<stdin>', 27), Token('LEFT_SQUARE', '[', '<stdin>', 27), Token('NUMBER', '1', '<stdin>', 27), Token('COMMA', ',', '<stdin>', 27), Token('LEFT_SQUARE', '[', '<stdin>', 27), Token('NUMBER', '1', '<stdin>', 27), Token('COMMA', ',', '<stdin>', 27), Token('NUMBER', '2', '<stdin>', 27), Token('COMMA', ',', '<stdin>', 27), Token('LEFT_SQUARE', '[', '<stdin>', 27), Token('NUMBER', '2', '<stdin>', 27), Token('COMMA', ',', '<stdin>', 27), Token('NUMBER', '3', '<stdin>', 27), Token('RIGHT_SQUARE', ']', '<stdin>', 27), Token('RIGHT_SQUARE', ']', '<stdin>', 27), Token('RIGHT_SQUARE', ']', '<stdin>', 27), Token('RETURN', 'return', '<stdin>', 29), Token('SYMBOL', 'b', '<stdin>', 29), Token('RIGHT_BRACKET', ')', '<stdin>', 30), Token('SYMBOL', 'd', '<stdin>', 32), Token('EQUAL', '=', '<stdin>', 32), Token('SYMBOL', 'convert', '<stdin>', 32), Token('LEFT_BRACKET', '(', '<stdin>', 32), Token('SYMBOL', 'a', '<stdin>', 32), Token('LEFT_BRACKET', '(', '<stdin>', 32), Token('RIGHT_BRACKET', ')', '<stdin>', 32), Token('LEFT_BRACKET', '(', '<stdin>', 32), Token('RIGHT_BRACKET', ')', '<stdin>', 32), Token('LEFT_SQUARE', '[', '<stdin>', 32), Token('NUMBER', '1', '<stdin>', 32), Token('RIGHT_SQUARE', ']', '<stdin>', 32), Token('LEFT_SQUARE', '[', '<stdin>', 32), Token('NUMBER', '2', '<stdin>', 32), Token('RIGHT_SQUARE', ']', '<stdin>', 32), Token('LEFT_SQUARE', '[', '<stdin>', 32), Token('NUMBER', '0', '<stdin>', 32), Token('RIGHT_SQUARE', ']', '<stdin>', 32), Token('COMMA', ',', '<stdin>', 32), Token('TEXT', 'number', '<stdin>', 32), Token('RIGHT_BRACKET', ')', '<stdin>', 32), Token('SYMBOL', 'print', '<stdin>', 33), Token('LEFT_BRACKET', '(', '<stdin>', 33), Token('LEFT_SQUARE', '[', '<stdin>', 33), Token('NUMBER', '1', '<stdin>', 33), Token('COMMA', ',', '<stdin>', 33), Token('NUMBER', '2', '<stdin>', 33), Token('RIGHT_SQUARE', ']', '<stdin>', 33), Token('LEFT_SQUARE', '[', '<stdin>', 33), Token('NUMBER', '0', '<stdin>', 33), Token('RIGHT_SQUARE', ']', '<stdin>', 33), Token('RIGHT_BRACKET', ')', '<stdin>', 33), Token('SYMBOL', 'print', '<stdin>', 34), Token('LEFT_BRACKET', '(', '<stdin>', 34), Token('TEXT', 'asd', '<stdin>', 34), Token('LEFT_SQUARE', '[', '<stdin>', 34), Token('NUMBER', '1', '<stdin>', 34), Token('RIGHT_SQUARE', ']', '<stdin>', 34), Token('LEFT_SQUARE', '[', '<stdin>', 34), Token('NUMBER', '0', '<stdin>', 34), Token('RIGHT_SQUARE', ']', '<stdin>', 34), Token('RIGHT_BRACKET', ')', '<stdin>', 34), Token('EOF', None, '<stdin>', 34)]